subsys/net/ip/tcp.c

```diff
@@ -826,7 +828,22 @@ int net_tcp_queue_data(struct net_context *context, struct net_pkt *pkt)
 
 	net_stats_update_tcp_sent(net_pkt_iface(pkt), data_len);
 
+	return net_tcp_queue_pkt(context, pkt);
+}
+
+/* This function is the sole point of *adding* packets to tcp->sent_list,
+ * and should remain such.
+ */
+static int net_tcp_queue_pkt(struct net_context *context, struct net_pkt *pkt)
+{
+	unsigned int key;
+
+	/* slist is not thread-safe, can be interrupted by retry timer and
+	 * accessed concurrently, protect against that.
+	 */
+	key = irq_lock();
```

Comment:
Please protect sent_list using a mutex here and in other places in this file.

Reply:
> Please protect sent_list using a mutex here

Yes, but why? As I told in https://github.com/zephyrproject-rtos/zephyr/pull/8674 , we should consider various strategies of protecting against concurrent access, and choose the best one (and that may be different best strategies for protecting various data).

Here, as the code comment says, net_tcp_queue_pkt() is the sole point of where something gets added to sent_list, and that's the only operation which can happen based on user API request, i.e. which may happen from the main thread, which is preemptive thread. All other accesses happen only from delayed work handlers, which run in cooperative thread, so can't be preempted themselves. So, we need to protect only this addition to list, all other access to sent_list are already serialized among each other.

Next point is that sys_slist_append() is very lightweight, maybe 3-5 cycles. So, if we just disable IRQs with irq_lock(), that's the max interrupt jitter we introduce. Mutex lock/unlock operations for sure include irq_lock()/irq_unlock() too, so they for sure introduce latency too. In other words, with mutexes, we don't win on latency, but surely lose on overall complexity and performance.

> and in other places in this file.

Well, that's certainly out of scope of this patch. It's also not needed, if the above analysis holds. If you don't buy my arguments right away, then I'd rather remove any attempts at protection of sent_list from this patch, and resubmit as a separate patch afterwards, so we can have focused discussion on it.

Reply:
Lets remove the locking then from this PR and think this more in another PR. There is also #8674 that deals with net_context locking and is somewhat related to this one.

Reply:
> Lets remove the locking then from this PR

Ok, doing.